File: SOC/src/cc/game/soc/core/Trade.java
package cc.game.soc.core;

import cc.lib.utils.Reflector;

/**
 * 
 * @author Chris Caron
 * 
 */
public final class Trade extends Reflector<Trade> {
	
	static {
		addAllFields(Trade.class);
	}
	
	private Card type;
	private int	amount = 0;

	public Trade() {}

    public String toString() {
        return type + " X " + amount;
    }
    
    public Trade(Card card, int amount) {
		this.type = card;
		this.amount = amount;
	}
    
	public Trade(ICardType type, int amount) {
		this.type = new Card(type, CardStatus.USABLE);
		this.amount = amount;
	}

	public ICardType getType() {
	    return type.getCardType().dereferenceOrdinal(type.getTypeOrdinal());
	}
	
    public int getAmount() {
        return amount;
    }
    
}

File: SOC/src/cc/game/soc/core/Trade.kt
package cc.game.soc.core

import cc.lib.utils.Reflector

/**
 *
 * @author Chris Caron
 */
class Trade : Reflector<Trade> {
	companion object {
		init {
			addAllFields(Trade::class.java)
		}
	}

	private val type: Card
	var amount = 0
		private set

	constructor() : this(ResourceType.Brick, 0)

	override fun toString(): String {
		return "$type X $amount"
	}

	constructor(card: Card, amount: Int) {
		type = card
		this.amount = amount
	}

	constructor(type: ICardType<*>, amount: Int) {
		this.type = Card(type, CardStatus.USABLE)
		this.amount = amount
	}

	fun getType(): ICardType<*> {
		return type.cardType.dereferenceOrdinal(type.typeOrdinal)
	}
}